components/script/dom/notification.rs

```diff
@@ -0,0 +1,444 @@
+    /// <https://notifications.spec.whatwg.org/#concept-language>
+    lang: DOMString,
+    /// <https://notifications.spec.whatwg.org/#silent-preference-flag>
+    silent: Option<bool>,
+    /// <https://notifications.spec.whatwg.org/#tag>
+    tag: DOMString,
+    /// <https://notifications.spec.whatwg.org/#concept-origin>
+    #[no_trace] // ImmutableOrigin is not traceable
+    origin: ImmutableOrigin,
+    /// <https://notifications.spec.whatwg.org/#vibration-pattern>
+    vibration_pattern: Vec<u32>,
+    /// <https://notifications.spec.whatwg.org/#timestamp>
+    timestamp: u64,
+    /// <https://notifications.spec.whatwg.org/#renotify-preference-flag>
+    renotify: bool,
+    /// <https://notifications.spec.whatwg.org/#require-interaction-preference-flag>
+    require_interaction: bool,
+    /// <https://notifications.spec.whatwg.org/#actions>
+    #[ignore_malloc_size_of = "NotificationAction"]
+    actions: Vec<NotificationAction>,
+    // TODO: image resource, icon resource, and badge resource
+}
+
+impl Notification {
+    #[cfg_attr(crown, allow(crown::unrooted_must_root))]
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn new(
+        global: &GlobalScope,
+        proto: Option<HandleObject>,
+        can_gc: CanGc,
+        title: DOMString,
+        options: RootedTraceableBox<NotificationOptions>,
+        origin: Option<ImmutableOrigin>,
+        base_url: Option<ServoUrl>,
+        fallback_timestamp: u64,
+    ) -> Fallible<DomRoot<Self>> {
+        let notification = Notification::new_inherited(
+            global,
+            title,
+            options,
+            origin,
+            base_url,
+            fallback_timestamp,
+        )?;
+        Ok(reflect_dom_object_with_proto(
+            Box::new(notification),
+            global,
+            proto,
+            can_gc,
+        ))
+    }
+
+    /// <https://notifications.spec.whatwg.org/#create-a-notification-with-a-settings-object>
+    fn new_inherited(
+        global: &GlobalScope,
+        title: DOMString,
+        options: RootedTraceableBox<NotificationOptions>,
+        origin: Option<ImmutableOrigin>,
+        base_url: Option<ServoUrl>,
+        fallback_timestamp: u64,
+    ) -> Fallible<Self> {
+        // If options["silent"] is true and options["vibrate"] exists, then throw a TypeError.
+        if options.silent.is_some() && options.vibrate.is_some() {
+            return Err(Error::Type(
+                "Can't specify vibration patterns when setting notification to silent.".to_string(),
+            ));
+        }
+        // If options["renotify"] is true and options["tag"] is the empty string, then throw a TypeError.
+        if options.renotify && options.tag.is_empty() {
+            return Err(Error::Type(
+                "tag must be set to renotify as an existing notification.".to_string(),
+            ));
+        }
+        // Set notification’s data to StructuredSerializeForStorage(options["data"]).
```

Comment:
You are missing a call to `https://html.spec.whatwg.org/multipage/structured-data.html#structuredserializeforstorage`, which may be found in [`dom/bindings/structuredclone` ](https://github.com/servo/servo/blob/0af1204aa3db47c80a50612491a084de859a5be1/components/script/dom/bindings/structuredclone.rs) and othewise would be a good todo.